src/lib/FreeC/Pass/KindInferencePass.hs

```diff
@@ -0,0 +1,211 @@
+
+-- | This module contains a compiler pass that checks if all type constructors
+--   in type expressions are fully applied and type variables are not applied.
+--
+--   For simplification, all type variables have kind @*@. Therefore, all type
+--   constructors must be fully applied, so partial application of type
+--   constructors is not allowed. This pass also checks that type variables and
+--   function types do not occur on the left hand side of a type application.
+--   If a type expression is found that does not match this rules, a fatal
+--   error is thrown. For simplification, a type expression that satisfies all
+--   of these rules is called kind-correct.
+--
+--   = Example
+--
+--   The following module contains three invalid definitions
+--
+--   > module Failures where
+--   >
+--   > data MyList a = MyNil | MyCons a (MyList a)
+--   >
+--   > fail1 :: MyList (MyList a a) -> Integer
+--   > fail1 x = 42
+--   >
+--   > fail2 :: a a -> Integer
+--   > fail2 x = 42
+--   >
+--   > fail3 :: MyList -> Integer
+--   > fail3 x = 42
+--
+--   = Specification
+--
+--   == Preconditions
+--
+--   The environment must contain all type declarations that are used in the
+--   module
```

Comment:
The environment does not contain declarations but entries for declarations.

```suggestion
--   The environment must contain entries for all type constructors that are
--   used in the module.
```

End every sentence with a period in the comments and error messages below as well.